Sources/PointFree/Session.swift

```diff
@@ -50,11 +50,122 @@ extension URLRequest {
+  public static let empty = Session(flash: nil, user: nil)
+
+  public var userId: Models.User.Id? {
+    get {
+      switch self.user {
+      case let .some(.ghosting(ghosteeId, _)):
+        return ghosteeId
+      case let .some(.standard(userId)):
+        return userId
+      case .none:
+        return nil
+      }
+    }
+  }
+
+  public var ghosterId: Models.User.Id? {
+    switch self.user {
+    case let .some(.ghosting(ghosteeId: _, ghosterId: ghosterId)):
+      return ghosterId
+    case .some(.standard), .none:
+      return nil
+    }
+  }
+
+  public var ghosteeId: Models.User.Id? {
+    switch self.user {
+    case let .some(.ghosting(ghosteeId: ghosteeId, ghosterId: _)):
+      return ghosteeId
+    case .some(.standard), .none:
+      return nil
+    }
+  }
+
+  public enum User: Codable, Equatable {
+    case ghosting(ghosteeId: Models.User.Id, ghosterId: Models.User.Id)
+    case standard(Models.User.Id)
+
+    private enum CodingKeys: CodingKey {
+      case ghosteeId
+      case ghosterId
+      case userId
+    }
+
+    public func encode(to encoder: Encoder) throws {
+      var container = encoder.container(keyedBy: CodingKeys.self)
+
+      switch self {
+      case let .ghosting(ghosteeId, ghosterId):
+        try container.encode(ghosteeId, forKey: .ghosteeId)
+        try container.encode(ghosterId, forKey: .ghosterId)
+
+      case let .standard(userId):
+        try container.encode(userId, forKey: .userId)
+      }
+    }
+
+    public init(from decoder: Decoder) throws {
+      let container = try decoder.container(keyedBy: CodingKeys.self)
+      do {
+        self = .standard(try container.decode(Models.User.Id.self, forKey: .userId))
+      } catch {
+        self = .ghosting(
+          ghosteeId: try container.decode(Models.User.Id.self, forKey: .ghosteeId),
+          ghosterId: try container.decode(Models.User.Id.self, forKey: .ghosterId)
+        )
+      }
+    }
+  }
+}
+
+extension Session: Codable {
+  private enum CodingKeys: CodingKey {
+    case flash
+    case user
+    case userId
+  }
+
+  public func encode(to encoder: Encoder) throws {
+    var container = encoder.container(keyedBy: CodingKeys.self)
+    try container.encodeIfPresent(self.flash, forKey: .flash)
+    switch self.user {
+    case let .some(.standard(userId)):
+      try container.encode(userId, forKey: .userId)
+    case .some(.ghosting):
+      try container.encodeIfPresent(self.user, forKey: .user)
+    case .none:
+      break
+    }
+  }
+
+  public init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+
+    self.flash = try container.decodeIfPresent(Flash.self, forKey: .flash)
+    do {
+      self.user = .standard(try container.decode(Models.User.Id.self, forKey: .userId))
+    } catch {
+      do {
+        self.user = try container.decode(Session.User.self, forKey: .user)
+      } catch {
+        self.flash = nil
+        self.user = nil
```

Comment:
we need this last catch all to make sure we get a decoded session, which can just be `nil` for both fields.

Reply:
Since we don't care about the error what if we did:
``` swift
self.user = (try? container.decode(Models.User.Id.self, forKey: .userId)).map(User.standard)
  ?? (try container.decode(Session.User.self, forKey: .user))
```

Reply:
👍 done